service/matching/ratelimit_manager.go

```diff
@@ -0,0 +1,188 @@
+
+	var (
+		effectiveRPS    float64
+		rateLimitSource enumspb.RateLimitSource
+	)
+	// Overall system rate limit will be the min of the two configs that are partition wise times the number of partions.
+	systemRPS := min(
+		r.config.AdminNamespaceTaskQueueToPartitionDispatchRate(),
+		r.config.AdminNamespaceToPartitionDispatchRate(),
+	) * r.getNumberOfReadPartitions()
+	r.systemRPS = systemRPS
+	switch {
+	case r.apiConfigRPS != nil:
+		effectiveRPS = *r.apiConfigRPS
+		rateLimitSource = enumspb.RATE_LIMIT_SOURCE_API
+	case r.workerRPS != nil:
+		effectiveRPS = *r.workerRPS
+		rateLimitSource = enumspb.RATE_LIMIT_SOURCE_WORKER
+	}
+
+	if effectiveRPS < r.systemRPS {
+		r.effectiveRPS = effectiveRPS
+		r.rateLimitSource = rateLimitSource
+	} else {
+		r.effectiveRPS = r.systemRPS
+		r.rateLimitSource = enumspb.RATE_LIMIT_SOURCE_SYSTEM
+	}
+}
+
+// Lazy injection of poll metadata.
+// Internally call UpdateRateLimit to share the same mutex.
+func (r *rateLimitManager) InjectWorkerRPS(meta *pollMetadata) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	var rps *float64
+	if meta != nil && meta.taskQueueMetadata != nil {
+		if workerRPS := meta.taskQueueMetadata.GetMaxTasksPerSecond(); workerRPS != nil {
+			value := workerRPS.GetValue()
+			rps = &value
+		}
+	}
+	r.workerRPS = rps
+	// updateRatelimitLocked includes internal logic to determine if an update is needed,
+	// so calling it unconditionally is safe and avoids redundant updates.
+	r.updateRatelimitLocked()
+}
+
+// Return the effective RPS and its source together.
+// The source can be API, worker or system default.
+// It defaults to the system dynamic config.
+func (r *rateLimitManager) GetEffectiveRPSAndSource() (float64, enumspb.RateLimitSource) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.computeEffectiveRPSAndSourceLocked()
+	return r.effectiveRPS, r.rateLimitSource
+}
+
+func (r *rateLimitManager) GetRateLimiter() quotas.RateLimiter {
+	return r.dynamicRateLimiter
+}
+
+// Updates the API-configured RPS based on the latest user data
+// and applies the new rate limit if the effective RPS has changed.
+func (r *rateLimitManager) UserDataChanged() {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	// Immediately recompute and apply rate limit if necessary.
+	r.updateRatelimitLocked()
+}
+
+// trySetRPSFromUserDataLocked sets the apiConfigRPS from user data.
+func (r *rateLimitManager) trySetRPSFromUserDataLocked() {
+	userData, _, err := r.userDataManager.GetUserData()
+	if err != nil {
+		return
+	}
+	config := userData.GetData().GetPerType()[int32(r.taskQueueType)].GetConfig()
+	if config == nil {
+		return
+	}
+	// If rate limit is an empty message, it means rate limit could have been unset via API.
+	// In this case, the apiConfigRPS will need to be unset.
+	rateLimit := config.GetQueueRateLimit()
+	if rateLimit.GetRateLimit() == nil {
+		r.apiConfigRPS = nil
+		return
+	}
+	val := float64(rateLimit.GetRateLimit().GetRequestsPerSecond())
+	r.apiConfigRPS = &val
```

Comment:
maybe we could do this to avoid an allocation each time?
```suggestion
	r.apiConfigRPS = &rateLimit.RateLimit.RequestsPerSecond
```
if it's a float64/32 thing, maybe change the types in here? we don't need even 32 bits of precision for any of this

Reply:
The dynamic configs are in Float64 and the dynamic rate limiter implementation also uses float64. Changing it will involve changing everything to float32. Do we want to do this?
This is why I was considering making changes in the api repo to use double for the request per seconds type to ensure everything is float64.
We will need to do this typecasting as long as we don't confirm to float64/float32 for all requests per second field. So I was thinking of keeping this as it is.

Reply:
ok, it's fine for now. we can try to clean it up later.